### PR TITLE
sql/sqltool: wrap flyway dir with fs

### PR DESCRIPTION
**Pull request by datdao**
### Extending FlywayDir Functionality with fs.FS

By wrapping `FlywayDir` with `fs.FS`, we can extend its functionality to include local directories and directories in memory or remote directories.

### CODE REVIEW

File: sql/sqltool/tool_test.go [hunk 1]
```diff
@@ -313,11 +314,22 @@ func TestChecksum(t *testing.T) {
...
 				"V3__third_migration.sql",
+				"v3/V3_1__fourth_migration.sql",
+				"R__views.sql",
+			},
+		},
+		{
+			name: "flyway non-local directory",
+			dir: func() migrate.Dir {
+				fs := fstest.MapFS{
+					"testdata/V1_initial.sql": &fstest.MapFile{Data: []byte("testdata")},
```
Comment (giautm) [inline sql/sqltool/tool_test.go:326]: please add more files, like previously tests

**Review by giautm — APPROVED** [review]
LGTM, @masseelch wdyt?

File: sql/sqltool/tool.go [hunk 2]
```diff
@@ -311,12 +312,16 @@ func NewFlywayDir(path string) (*FlywayDir, error) {
...
 			return err
 		}
-		if path != "" && e.IsDir() {
-			h, err := hidden(filepath.Join(d.Path(), path))
+		if path != "." && e.IsDir() {
+			fullPath := path
+			if d, ok := d.FS.(*migrate.LocalDir); ok {
+				fullPath = filepath.Join(d.Path(), path)
+			}
+			h, err := hidden(fullPath)
```
Comment (a8m) [inline sql/sqltool/tool.go:324]:
```suggestion
			fullPath := path
			if l, ok := d.FS.(*migrate.LocalDir); ok {
				fullPath = filepath.Join(l.Path(), path)
			}
			h, err := hidden(fullPath)
```
Don't shadow the receiver

File: sql/sqltool/tool.go [hunk 3]
```diff
@@ -292,7 +293,7 @@ func (f *DBMateFile) Stmts() ([]string, error) {
 type (
 	// FlywayDir wraps migrate.LocalDir and provides a migrate.Scanner implementation able to understand files
 	// generated by the FlywayFormatter for migration directory replaying.
-	FlywayDir struct{ *migrate.LocalDir }
+	FlywayDir struct{ fs.FS }
```
Comment (masseelch) [inline sql/sqltool/tool.go:296]: Edit comment as well

File: sql/sqltool/tool.go [hunk 4]
```diff
@@ -351,6 +356,23 @@ func (d *FlywayDir) Files() ([]migrate.File, error) {
...
+}
+
+// WriteFile implements Dir.WriteFile.
+func (d *FlywayDir) WriteFile(name string, b []byte) error {
+	if d, ok := d.FS.(*migrate.LocalDir); ok {
+		return d.WriteFile(name, b)
+	}
+	return errors.New("sql/sqltool: cannot write to non-local directory")
+}
+
```
Comment (masseelch) [inline sql/sqltool/tool.go:375]: Wrapping Dir means things like `MemDir` work for this as well.

**Comment by giautm** [comment]
> Like the idea. However, I suggest to wrap `migrate.Dir` instead of `fs.FS`.

Because `migrate.Dir` has more method than `fs.FS` so we can't use FlywayDir with `os.DirFS` or other file-system.

File: sql/sqltool/tool.go [hunk 5]
```diff
@@ -351,6 +356,23 @@ func (d *FlywayDir) Files() ([]migrate.File, error) {
 	return ret, nil
 }
 
+// Checksum implements Dir.Checksum. By default, it calls Files() and creates a checksum from them.
+func (d *FlywayDir) Checksum() (migrate.HashFile, error) {
```
Comment (giautm) [inline sql/sqltool/tool.go:360]:
```suggestion
func (d *FlywayDir) Checksum() (migrate.HashFile, error) {
	if check, ok d.FS.(migrate.Dir); ok {
		return check.Checksum()
	}
```

File: sql/sqltool/tool.go [hunk 6]
```diff
@@ -311,12 +312,16 @@ func NewFlywayDir(path string) (*FlywayDir, error) {
...
-	if err := fs.WalkDir(d, "", func(path string, e fs.DirEntry, err error) error {
+	if err := fs.WalkDir(d, ".", func(path string, e fs.DirEntry, err error) error {
 		if err != nil {
 			return err
 		}
-		if path != "" && e.IsDir() {
-			h, err := hidden(filepath.Join(d.Path(), path))
+		if path != "." && e.IsDir() {
+			fullPath := path
+			if l, ok := d.FS.(*migrate.LocalDir); ok {
```
Comment (giautm) [inline sql/sqltool/tool.go:321]:
```suggestion
			if l, ok := d.FS.(interface { Path() string }); ok {
```

File: sql/sqltool/tool.go (hunk 5 above)
Comment (masseelch) [reply sql/sqltool/tool.go:360]: ?

File: sql/sqltool/tool.go [hunk 7]
```diff
@@ -351,6 +356,26 @@ func (d *FlywayDir) Files() ([]migrate.File, error) {
...
+	}
+	return migrate.NewHashFile(files)
+}
+
+// WriteFile implements Dir.WriteFile.
+func (d *FlywayDir) WriteFile(name string, b []byte) error {
+	if d, ok := d.FS.(migrate.Dir); ok {
+		return d.WriteFile(name, b)
+	}
+	return errors.New("sql/sqltool: cannot write to non-local directory")
```
Comment (masseelch) [inline sql/sqltool/tool.go:376]:
```suggestion
	return errors.New("sql/sqltool: write not supported")
```
This is more accurate, isn't it?

File: sql/sqltool/tool_test.go [hunk 8]
```diff
@@ -320,6 +322,27 @@ func TestChecksum(t *testing.T) {
...
+			dir: func() migrate.Dir {
+				fs := fstest.MapFS{
+					"U1__initial.sql":               &fstest.MapFile{Data: []byte("U1__initial")},
+					"V1__initial.sql":               &fstest.MapFile{Data: []byte("V1__initial")},
+					"R__views.sql":                  &fstest.MapFile{Data: []byte("R__views")},
+					"B2__baseline.sql":              &fstest.MapFile{Data: []byte("B2__baseline")},
+					"V2__second_migration.sql":      &fstest.MapFile{Data: []byte("V2__second_migration")},
+					"V3__third_migration.sql":       &fstest.MapFile{Data: []byte("V3__third_migration")},
+					"v3/V3_1__fourth_migration.sql": &fstest.MapFile{Data: []byte("V3__1_fourth_migration")},
+				}
```
Comment (masseelch) [inline sql/sqltool/tool_test.go:336]: <3

**Review by masseelch — APPROVED** [review]

File: sql/sqltool/tool.go (hunk 5 above)
Comment (giautm) [reply sql/sqltool/tool.go:360]: respect the checksum implement in .FS if existing

**Review by giautm — APPROVED** [review]

**Pull request merged**